Description: Yate.pm: full control over message's reponses
Author: vir@ctm.ru
---
This patch allows perl handler to modify message's retval without acknoledging message reception.
This can be user in engine command completion and in several other cases.
Index: yate/share/scripts/Yate.pm
===================================================================
--- yate.orig/share/scripts/Yate.pm	2016-01-15 16:52:18.000000000 +0300
+++ yate/share/scripts/Yate.pm	2016-01-15 16:53:29.000000000 +0300
@@ -37,7 +37,7 @@
     use Data::Dumper;
 
     # Set version && disable output buffering.
-    our $VERSION = '0.22';
+    our $VERSION = '0.25';
     $ |= 1;
 }
 
@@ -522,7 +522,11 @@
     foreach (@{$self->{'_handlers'}->{$self->header('name')}}) {
 	my $return = $_->($self);
 
-	if (defined($return) && lc($return) ne 'false' && $return ne '0') {
+	if(ref($return) eq 'ARRAY') {
+	    $self->error('Invalid array returned from ' . $self->header('name') . ' event handler') unless @$return == 2;
+	    $self->return_message(@$return);
+	    return 1;
+	} elsif (defined($return) && lc($return) ne 'false' && $return ne '0') {
 	    $self->return_message('true', $return);
 
 	    return 1;
@@ -827,6 +831,11 @@
 wish to process a message, but not to send anything back to the Engine
 return ''.
 
+If you want full control over message's return values, uou can use
+array reference as your handler return value. Array must contain
+exactly two elements: 'handled' value ('true' or 'false') and
+message 'retvalue' text.
+
 Handlers will have only one argument and that is the current Yate
 object itself. You can access headers and parameters by C<header> and
 C<param>. For debugging the current structure of the object use C<dump>.
